UMS/doc/man/man3/updateMachine.3
'\" t
.\"     Title: updateMachine
.\"    Author:  SysFera SA <info@sysfera.com>
.\" Generator: DocBook XSL Stylesheets v1.78.0 <http://docbook.sf.net/>
.\"      Date: May 2014
.\"    Manual: UMS C++ API Reference
.\"    Source: VISHNU 4.0.0 alpha
.\"  Language: English
.\"
.TH "UPDATEMACHINE" "3" "May 2014" "VISHNU 4.0.0 alpha" "UMS C++ API Reference"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
updateMachine \- updates machine description
.SH "SYNOPSIS"
.HP \w'int\ vishnu::updateMachine('u
.BI "int vishnu::updateMachine(const\ string&\ " "sessionKey" ", const\ Machine&\ " "machine" ");"
.SH "DESCRIPTION"
.PP
This command allows an admin to update a VISHNU machine or to locked it\&. A machine locked is not usable\&.
.SH "ARGUMENTS"
.PP
\fIsessionKey\fR
.RS 4
Input argument\&. The sessionKey is the encrypted identifier of the session generated by VISHNU\&.
.RE
.PP
\fImachine\fR
.RS 4
Input argument\&. Existing machine information\&.
.RE
.SH "EXCEPTIONS"
.PP
The following exceptions may be thrown:
.PP
"Vishnu not available (Service bus failure)" [1]
.RS 4
.RE
.PP
"Vishnu not available (Database error)" [2]
.RS 4
.RE
.PP
"Vishnu not available (Database connection)" [3]
.RS 4
.RE
.PP
"Vishnu not available (System)" [4]
.RS 4
.RE
.PP
"Internal Error: Undefined exception" [9]
.RS 4
.RE
.PP
"The user is not an administrator" [25]
.RS 4
.RE
.PP
"The session key is unrecognized" [28]
.RS 4
.RE
.PP
"The sessionKey is expired\&. The session is closed\&." [29]
.RS 4
.RE
.PP
"The machine id is unknown" [32]
.RS 4
.RE
.PP
"The closure policy is unknown" [42]
.RS 4
.RE
.SH "AUTHOR"
.PP
\fB SysFera SA\fR <\&info@sysfera.com\&>
.RS 4
Developed the first version of these man pages as part of the VISHNU project.
.RE
.SH "COPYRIGHT"
.br
Copyright \(co 2013 SysFera SA
.br
.PP
These manual pages are provided under the following conditions:
.PP
Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  1." 4.2
.\}
Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.sp -1
.IP "  2." 4.2
.\}
Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
.RE
.PP
This software is governed by the CECILL licence under French law and abiding by the rules of distribution of free software. You can use, modify and/ or redistribute the software under the terms of the CeCILL license as circulated by CEA, CNRS and INRIA at the following URL "http://www.cecill.info".
.PP
As a counterpart to the access to the source code and rights to copy, modify and redistribute granted by the license, users are provided only with a limited warranty and the software's author, the holder of the economic rights, and the successive licensors have only limited liability.
.PP
In this respect, the user's attention is drawn to the risks associated with loading, using, modifying and/or developing or reproducing the software by the user in light of its specific status of free software, that may mean that it is complicated to manipulate, and that also therefore means that it is reserved for developers and experienced professionals having in-depth computer knowledge. Users are therefore encouraged to load and test the software's suitability as regards their requirements in conditions enabling the security of their systems and/or data to be ensured and, more generally, to use and operate it in the same conditions as regards security.
.sp


ChangeLog
+----------------------+
| VISHNU 4.0.0 - alpha |
+----------------------+
* New architecture : one binary for all servers "xmssed"
* New architecture for file transfers, when the client machine is in the transfer, it launches the transfer.
* IMS module removed, UMS module gets export and definition of identifiers
* TMS module API has been cleaned (no more machineId necessary when not needed)
* New services : vishnu_ping, vishnu_list_server, vishnu_get_server_version
* Messages are in JSON (some emf serialized object still remain in some JSON messages)
* New configuration keys (less URIs, servers to register or not in the dispatcher, TMS user mode to use)
* TMS can change the user using either SSH or setuid based on the configuration of the platform
* Process start writing in syslog
* Add native OpenNebula support


+--------------+
| VISHNU 3.2.1 |
+--------------+
* Fix ticket #359, #364, #345, #333, #382, #337, #403, #409, #392, #350
* AuthSystem names can contain the 'name' variable now

+--------------+
| VISHNU 3.2.0 |
+--------------+
* Added compatibility with OpenNebula
* Support swig2 and python 2.7 (ticket 384)
* Fix ticket 379, 380, 387, 330, 331

+--------------+
| VISHNU 3.1.1 |
+--------------+
* Move IMS define_*_identifier functions to UMS module
* Update the algorithm to generate identifiers, counters are not mandatory, if an existing login is generated, a unique number will be added at the end of the login
* Improve serialisation and message handling
* Fix bug when starting dispatcher with invalid default file parameter
* Clean documentation (no WSDL mention anymore)

+--------------+
| VISHNU 3.1.0 |
+--------------+
* Added support for SSL communication with DB and zmq
* New mechanism to externalize SQL request

+--------------+
| VISHNU 3.0.0 |
+--------------+
* Fixed a bug with cancel (tms), *cp, *mv (fms)
* Remove deprecated parameters from account and machine (ssh keys)
* Remove database sql delete, now a status deleted is used
* Fixed a few bugs: #348, #414, #424, #454
* Option VISHNU_USE_LDAP in cmake has been renamed ENABLE_LDAP to be more
	consistent with the other options (ENABLE_JAVA, ENABLE_MYSQL,
	ENABLE_POSTGRESQL, ENABLE_PYTHON, ENABLE_SWIG)

+----------------------+
| VISHNU 3.0.0 - Beta2 |
+----------------------+
* Added beta support for submitting jobs to a Cloud, using DeltaCloud
  (tested with OpenStack only). Currently, only the Vishnu user is
  used, there is no user switch during submission. The credentials
  need to be provided to VISHNU though an configuration file,
  options, or environment variables.
* /!\ The database needs to be updated using scripts in
  core/database/database_update_addvminfo_*.sql
* /!\ The compilation process has been updated: you can now compile the
  client CLI and the servers separately
* Automtests have been updated: they do not launch the Vishnu platform,
  they use an already deployed one.
* fixed a bug where signal-interrupted zmq calls (EINTR) were not
  handled properly
* fixed a bug on the "POSIX" backend where the wallclock time was not
  correctly handled which caused submission scripts to sometimes be
  deleted before being executed.
* Files necessary to build vishnu java library have been included in this
  repository. Currently there is no doc for this, see
  scripts/ALL_prepareScript.sh if you want the lib.

+----------------------+
| VISHNU 3.0.0 - Beta1 |
+----------------------+
* Added sample configuration files for VISHNU and Supervisord
* Fixed several memory leaks
* Fixed bad boost version in CMakeLists
* Unified configuration files: all Vishnu daemons and clients can have the
	same configuration file
* New documentation user/admin
* Add posix-shell: you can start TMS without batch scheduler, TMS functionalities available in the shell
* Fixed bugzilla #341 - Rename of FMS commands using POSIX name
* WS are abandoned, they won't be maintained anymore. Related files and
	directories have been removed.
* Java APIs are available
* Vishnu does not rely on DIET anymore. All communications are handled by
	ZMQ. SeDs only require 1 open port to communicate (easier for
	deployments)
* TMS now uses FMS to send input files to jobs. You can use remote files
	as input files.
* A new daemon has been introduced: the dispatcher which serves as a proxy
	between clients and SeDs
* Several bugs closed: #449, #452, #453, #225, #409, #344, #291, #289,
	#288, #226, #292, #295, #441, #233, #341, #386, #447, #436, #446

+-----------------------------+
| VISHNU 2.0.0 - Beta3        |
+-----------------------------+
* Improve the generation of the ID for Vishnu objects
* Add documentation for admin with distant databases
* Clean unused files
* Fixed bug 440
* Fix LSF bug for -N and -P options in submit
* Fix bad types in comparison from unsigned int to size_t
* Add PBS_Pro support
* Add specific parameters for batchs schedulers (option -S with vishnu_submit_job)
* Add default parameters in config files for batchs
* WS do not work

+-----------------------------+
| VISHNU 2.0.0 - Beta2        |
+-----------------------------+

* Fixed bugzilla #417 - compilation problems with LoadLeveler
* Fixed bugzilla #410 - solve restoreConfiguration service problems when webBoard data have been saved in the database
* Fixed bugzilla #431 - vishnu_get_job_output renvoie une erreur Dagda
* Fixed bugzilla #426 alias #433 - vishnu_list_jobs with a list of statuses failed (vishnu_list_job -S)
* Fixed bugzilla #434 - vishnu_get_job_output throws core dump exception when the error path or the output do not longer exist
* Fixed error to try to reconnect when having mysql error server has gone
* Changed - input files are now stored in the local account's home directory

+-----------------------------+
| VISHNU 2.0.0 - Beta version |
+-----------------------------+

* Added status display on textual form for vishnu_list_users and vishnu_list_machines 
* Added Client/Server version management for services: connect, reconnect and changePassword

UMS/src/cli/updateMachine.cc


#include "updateMachine.hh"
#include "utils.hh"
#include "machineUtils.hpp"
#include <boost/bind.hpp>

namespace po = boost::program_options;

using namespace std;

int main (int ac, char* av[]){




		/******* Parsed value containers ****************/

		string dietConfig;

		std::string name;

		std::string site;

		std::string language;

		std::string machineDescription;

		std::string machineId;

		std::string sessionKey;


		/********** EMF data ************/

		UMS_Data::Machine upMachine;



/**************** Describe options *************/

		boost::function1<void,string> fName( boost::bind(&UMS_Data::Machine::setName,boost::ref(upMachine),_1));

		boost::function1<void,string> fSite( boost::bind(&UMS_Data::Machine::setSite,boost::ref(upMachine),_1));

			boost::function1<void,string> fLanguage( boost::bind(&UMS_Data::Machine::setLanguage,boost::ref(upMachine),_1));


			boost::shared_ptr<Options> opt= makeMachineOptions(av[0], fName,dietConfig, fSite,fLanguage);

        opt->add("sessionKey",
                        "The session key",
                        ENV,
                        sessionKey);

				opt->add("machineId",
											"The identifier of the machine",
												HIDDEN,
												machineId,
												1);

				opt->setPosition("machineId",-1);

				opt->add("machineDescription,d",
											"The description of the machine",
												CONFIG,
												machineDescription);

        try{


/**************  Parse to retrieve option values  ********************/

		opt->parse_cli(ac,av);

		opt->parse_env(env_name_mapper());

		opt->notify();





/********  Process **************************/

		if (opt->count("machineId")){

			cout <<"The machine identifier is " << machineId << endl;

		}

		if (opt->count("name")){

			cout <<"The name of the machine is " <<  upMachine.getName() << endl;

		}

		if(opt->count("site")){

			cout <<"the site is : " << upMachine.getSite() << endl;

		}

		if(opt->count("language")){

			cout << "The language is " <<  upMachine.getMachineId() << endl;

		}

		if(opt->count("machineDescription")){//Fix me

		cout << "Enter the Machine Description:\n ";


		getline(cin, machineDescription);

		upMachine.setMachineDescription(machineDescription);

		}


		checkVishnuConfig(*opt);

/************** Call UMS connect service *******************************/

               // initializing DIET

							  if (diet_initialize(dietConfig.c_str(), ac, av)) {
                    cerr << "DIET initialization failed !" << endl;
               return 1;
              }

							 updateMachine(sessionKey,upMachine);

	}// End of try bloc

				catch(po::required_option& e){// a required parameter is missing

					usage(*opt," machineId ","required parameter is missing");
				}

				catch(std::exception& e){

					errorUsage(av[0],e.what());

					return 1;
  }

  return 0;

}// end of main


